crates/asm/src/state.rs

```diff
@@ -0,0 +1,96 @@
+pub struct AnchorState {
+    /// The current view of the L1 chain required for state transitions.
+    pub chain_view: ChainViewState,
+
+    /// States for each subprotocol section, sorted by Subprotocol Version/ID.
+    pub sections: Vec<SectionState>,
+}
+
+/// Represents the on‐chain view required by the Anchor State Machine (ASM) to process
+/// state transitions for each new L1 block.
+#[derive(Debug, Clone)]
+pub struct ChainViewState {
+    /// All data needed to validate a Bitcoin block header, including past‐n timestamps,
+    /// accumulated work, and difficulty adjustments.
+    pub pow_state: HeaderVerificationState,
+
+    /// Events emitted by subprotocols, keyed by L1 block:
+    /// - The outer Vec is ordered by block ID.
+    /// - Each inner Vec contains tuples of `(subprotocol_id, event_hash)`.
+    ///
+    /// `subprotocol_id` (u8) identifies which subprotocol emitted the event,
+    /// and `Buf32` is the protocol‐computed hash of the event payload.
+    // TODO: Eventually this will use an MMR for minimal, non‐linear‐growing on‐chain state.
+    pub events: Vec<(L1BlockId, Vec<(u8, Buf32)>)>,
+}
+
+/// Holds the off‐chain serialized state for a single subprotocol section within the ASM.
+///
+/// Each `SectionState` pairs the subprotocol’s unique ID with its current serialized state,
+/// allowing the ASM to apply the appropriate state transition logic for that subprotocol.
+#[derive(Debug, Clone)]
+pub struct SectionState {
+    /// Identifier of the subprotocol
+    pub subprotocol_id: u8,
+
+    /// The serialized subprotocol state.
+    ///
+    /// This is normally fairly small, but we are setting a comfortable max limit.
+    pub data: Vec<u8>,
+}
+
+impl AnchorState {
+    /// Constructs a Vec of all successfully deserialized subprotocol instances
+    pub fn subprotocols(&self) -> Vec<impl Subprotocol> {
+        let mut protocols = Vec::with_capacity(self.sections.len());
+        for section in &self.sections {
+            // Ignore any sections that fail to deserialize
+            if let Ok(proto) = section.subprotocol() {
+                protocols.push(proto);
+            }
+        }
+        protocols
+    }
+}
+
+impl SectionState {
+    /// Returns the identifier of the subprotocol for this section.
+    ///
+    /// This ID corresponds to the version or namespace of the subprotocol whose
+    /// state is serialized in this section.
+    pub fn id(&self) -> u8 {
+        self.subprotocol_id
+    }
+
+    /// Attempts to deserialize this section into its corresponding subprotocol state machine.
+    ///
+    /// Uses the section’s `subprotocol_id` to select and construct the appropriate
+    /// `Subprotocol` implementation via `TryFrom<SectionState>`.
+    pub fn subprotocol(&self) -> Result<impl Subprotocol, ASMError> {
+        match self.id() {
+            CORE_SUBPROTOCOL_ID => CoreASMState::try_from(self),
+            _ => Err(ASMError::InvalidSubprotocol(self.id())),
+        }
```

Comment:
This not having a verb-y name implies that it's a simple accessor, but it's actually doing a fallible deserialization.  So on its own it should be `to_subproto_state`.

This is actually the source of the issue in the `subprotocols` function.  It's not manifesting here because there only is a single impl for `Subprotocol` that exists in the body of the function, but it would break if you tried to implement the bridge subproto.

It would make more sense for this to be defined as `fn try_parse_subproto_state<T: Subprotocol>(&self) -> Result<T::State, Error>` and *only* attempt deserializing as that one subproto's state type.